Outsource uncertainty

shawndove opened the pull request.

shawndove committed:
store raw cube data

shawndove committed:
calculate bootstrap CIs using dubicube and add them to calculated indicator_ts object

shawndove committed:
add dubicube as dependency

shawndove committed:
add documentation and indicator level CIs

shawndove committed:
export add_ci

shawndove committed:
remove ci calculations (offloaded to add_ci)

shawndove committed:
documentation update

shawndove committed:
first commit

shawndove committed:
version update

shawndove committed:
version update

shawndove committed:
add dubicube to remotes

shawndove committed:
fix dubicube github url

shawndove committed:
update CI calculation in vignette

shawndove committed:
remove leftover confidence interval code and adjust documentation

shawndove committed:
remove CI code artifacts

shawndove committed:
fix spelling mistake

shawndove committed:
documentation fix

shawndove committed:
add missing null assignment

shawndove committed:
fix documentation, add missing null assignments, and fix incorrectly named object

shawndove committed:
fix roxygen2 code style issue

shawndove committed:
fix documentation typo

shawndove committed:
add separate grouping logic for species occurrences and species range

shawndove committed:
prevent warning during testing of single point geometry

shawndove committed:
add missing parameter

shawndove committed:
fix broken tests

shawndove committed:
remove invalid paramter from documentation

shawndove committed:
increase hill diversity bootstrap defaults to 1000

shawndove committed:
documentation update

shawndove committed:
fix broken tests

shawndove commented:
This update moves the uncertainty calculation to a separate step of the workflow. Now all indicators except Hill diversity (which calculates internally using iNEXT) are calculated without CIs. The user can then add them using the add_ci() function, which defaults to bootstrapping the whole dataset using dubicube, but also provides the option to do it at the indicator level using the internal code.
Putting the uncertainty in a separate step was more work, but it should be more robust and easier to update. It will also make it easier to add ci function for maps in the future.

wlangera commented on R/add_ci.R:
```diff
@@ -0,0 +1,207 @@
+#' @param overwrite (Optional) Logical. If the indicator already contains
+#'   confidence intervals (`ll` and `ul` columns), should they
+#'   be replaced? (Default: TRUE)
+#' @param ... Additional arguments passed to the underlying functions.
```
Which underlying functions? I do not think it is used now?
Do you mean for the two dubicube functions?

In that case, you can perhaps use new arguments:
boot_args = list(): Named list of additional arguments passed to `dubicube::bootstrap_cube()`
ci_args = list(): Named list of additional arguments passed to `dubicube::calculate_bootstrap_ci()`

wlangera commented on R/add_ci.R:
```diff
@@ -0,0 +1,207 @@
+#'
+#' @details
+#' The function acts as a bridge to the `dubicube` package for statistical
+#' heavy lifting. For certain indicators (e.g., Hill numbers), confidence
```
For every indicator I would clearly say if it is calculated with bootstrapping or not (not "e.g. Hill numbers", but a full list). That way it is clear for the user and future developers.

wlangera reviewed (changes requested):
Hi Shawn,
Thanks for the quick work. I doubt I will have time to do a complete review in January, because of my other deliverables.
Therefore, I already made some comments so you can work further on this if you want.

We should also have unit tests and perhaps a vignette. This vignette could also just be conceptual with links to the dubicube package.

Let me know what you think

shawndove committed:
feat(uncertainty): Implement add_ci with dubicube integration and add vignette

shawndove committed:
conductor(plan): Mark Phase 2 as COMPLETED and Phase 3 as IN PROGRESS

shawndove committed:
fix(uncertainty): Fix full_join error in add_ci and include bootstrap summaries in get_bootstrap_ci

shawndove committed:
docs(vignette): Add dubicube exploration details and links

shawndove committed:
chore(metadata): Update documentation and project metadata for version 0.9.0

shawndove committed:
fix(docs): Standardize roxygen2 tags and ignore files to fix building and loading

shawndove committed:
conductor(plan): Update commit hashes for integration and documentation

shawndove committed:
fix(ci): Resolve R-CMD-check failures by fixing ignore file encoding and updating documentation

shawndove committed:
test(utils): Fix expected type in breaks_pretty_int for NA input

pretty(NA) returns numeric(0), not logical(0). This resolves a test failure
in R-devel where strict type checking caught the discrepancy.

shawndove committed:
fix type mismatch

shawndove committed:
refactor(ci): Correctly categorize indicators for whole-cube vs group-specific bootstrapping

- Species-level indicators (spec_occ, spec_range) now use group-specific logic.
- Aggregate indicators use whole-cube logic.
- Added critical guideline regarding manual ignore file management.

wlangera reviewed (commented):
I have updated dubicube.

Here is a review of the documentation.

wlangera committed:
add space

wlangera committed:
identify bootstrap method

wlangera committed:
prepare indicator bootstrapping with dubicube

wlangera committed:
add transformation functions

wlangera committed:
add function to namespace

wlangera committed:
add unit tests

wlangera committed:
fix bug

wlangera committed:
parse no_bias argument according to dubicube v0.11.0

wlangera committed:
update documentation

shawndove committed:
Merge pull request #103 from b-cubed-eu/update-dubicube-ward

Parse dubicube arguments

shawndove committed:
reorder ci_type options, and update documentation

shawndove committed:
update news items to reflect merged changes

shawndove committed:
update after merged changes

shawndove committed:
update to reflect merged changes

shawndove committed:
Merge branch 'main' into outsource_uncertainty

shawndove committed:
fix malformed function call

shawndove committed:
fix missing package issue for r-cmd-check

shawndove committed:
fix missing colon

shawndove committed:
add another missing package that fell off CRAN to remotes

shawndove committed:
fix missing comma

shawndove committed:
another attempt to fix missing dependency

shawndove committed:
another attempt to fix missing dependency

shawndove committed:
another attempt to fix missing dependency

shawndove committed:
another attempt to solve missing dependency

shawndove committed:
fix malformed function call

shawndove committed:
attempt r-cmd-check fix through workflow update

shawndove committed:
call dubicube as an extra package

shawndove committed:
try again

shawndove committed:
fix accidental rename of bootstrap_results

shawndove committed:
add ellipsis back to parameters, as it is used to transfer paramters to calc_ci. Also fix year having wrong type

shawndove committed:
documentation update

shawndove committed:
fix year to handle more type mismatches

shawndove committed:
use all_of with select to fix warning

shawndove committed:
add missing seed parameter

shawndove committed:
fix broken tests

shawndove committed:
only add seed to bootstrap_params if not NULL

shawndove committed:
fix select line

shawndove committed:
documentation update

shawndove committed:
fix parameter naming mismatch with dubicube

shawndove committed:
suppress 'extreme order statistics' warning due to low replicates

shawndove committed:
fix missing seed parameter

shawndove committed:
set default seed and adjust logit function to avoid failure when evenness is zero

shawndove committed:
reduce bootstraps to make faster

shawndove committed:
fix: create composite key for multiple grouping variables

dubicube only accepts a single grouping variable, but spec_occ and
spec_range indicators need to group by both year and taxonKey.

Solution: Create a composite grouping key by concatenating multiple
grouping variables with underscore separator before sending to dubicube.

Changes:
- Added logic to detect when multiple grouping columns are needed
- Creates composite key using paste/apply for robust concatenation
- Updates group_cols to use single "group_key" column
- Maintains original grouping semantics while satisfying dubicube API

Fixes bootstrapping failure for spec_occ and spec_range indicators.

shawndove committed:
fix: add validation for missing grouping columns

shawndove committed:
fix: preserve raw data without indicator classes for bootstrapping

shawndove committed:
fix: update test mock data to include required grouping columns

shawndove committed:
fix: update test expectations for new composite key behavior

shawndove committed:
fix: apply indicator class to data for S3 method dispatch during bootstrapping

shawndove committed:
fix: ensure indicator class is first for S3 dispatch and add debug output

shawndove committed:
test: add S3 dispatch verification test and diagnostic script

shawndove committed:
chore: remove temporary diagnostic script

shawndove committed:
Merge branch 'main' into outsource_uncertainty

shawndove committed:
documentation update

shawndove committed:
remove irrelevant test

shawndove committed:
update noci_list

shawndove committed:
fix class assignment, add reverse group split logic, force taxonKey to numeric

shawndove committed:
add temporary dubicube replacement fns used for local debugging

shawndove committed:
increase num_bootstraps

shawndove committed:
revert to original

shawndove committed:
Merge branch 'main' into outsource_uncertainty

wlangera commented:
I have implemented your suggestions. Indeed my package has no fail-safe for NULL or small group sizes. This is an open issue. Please work with [this version](https://github.com/b-cubed-eu/dubicube/pull/104):

```r
remotes::install_github("b-cubed-eu/dubicube#104)
```

I will merge it with the main once we successfully closed this PR

shawndove committed:
Merge branch 'main' into outsource_uncertainty

shawndove commented:
> I have implemented your suggestions. Indeed my package has no fail-safe for NULL or small group sizes. This is an open issue. Please work with [this version](https://github.com/b-cubed-eu/dubicube/pull/104):
> 
> ```r
> remotes::install_github("b-cubed-eu/dubicube#104)
> ```
> 
> I will merge it with the main once we successfully closed this PR

I should have this ready to go soon. I'm fixing the update to be compatible with the changes I've made in the last weeks.

shawndove committed:
remove hill diversity from noci_list

shawndove committed:
set default bootstraps to 0 for hill diversity so it does not calculate them on the main workflow

shawndove committed:
replace inherits with rlang version for stopifnot_error

shawndove committed:
add hill diversity to calc_ci methods

shawndove committed:
replace inherits with rlang version for stopifnot_error

shawndove committed:
replace inherits with rlang version for stopifnot_error

shawndove committed:
add default values if not present, and replace inherits with rlang version for stopifnot_error

shawndove committed:
add hill diversity

shawndove committed:
fix tests

shawndove committed:
Merge branch 'outsource_uncertainty' of https://github.com/b-cubed-eu/b3gbi into outsource_uncertainty

shawndove committed:
set defaults for ci_type and num_bootstrap none and 0 if missing

shawndove committed:
add null assignments for hill diversity

shawndove committed:
add area to null assignments

shawndove committed:
add hill diversity calc_ci methods

shawndove committed:
add hill diversity calc_ci methods

shawndove committed:
remove ci_type and num_bootstrap

shawndove committed:
update

shawndove committed:
add dot filtering and wrap calc functions in do.call

shawndove committed:
fix tests

shawndove committed:
add expected_years as a parameter for bootstrapping

shawndove committed:
cosmetic change to dot parameter extraction code

shawndove committed:
fix test

shawndove committed:
add expected years

shawndove committed:
add expected_years parameter to documentation

shawndove committed:
document expected_years

shawndove committed:
fix insufficient data warning

shawndove committed:
add unit tests to increase coverage

shawndove committed:
add extra unit tests to increase patch coverage

shawndove committed:
add extra test to increase patch coverage

shawndove committed:
change default ci method from norm to perc

shawndove committed:
change default ci method from norm to perc

shawndove committed:
cosmetic and functional adjustments

shawndove committed:
update to reflect percentile intervals as default

shawndove committed:
update to reflect percentile intervals as default and explain types of intervals

shawndove committed:
update warning about hill diversity bootstrapping

shawndove committed:
add fast paths for single year and single group bootstrap calls

shawndove committed:
update spec_range and spec_occ to use whole cube bootstrapping

shawndove committed:
fix broken tests

shawndove committed:
update

shawndove committed:
update

shawndove committed:
update one file size

shawndove committed:
remove mention of bugs that were both introduced and fixed during the creation of this update

shawndove committed:
fix warning about deprecated use of external vector

wlangera reviewed (approved):
Thanks Shawn!
Is it intended that all chunks in the uncertainty vignette are `eval=FALSE`?

shawndove committed:
Merge branch 'fix-package-bugs-on-main' into outsource_uncertainty

shawndove committed:
test(hex): prevent ISEA3H native fallback failures on platforms with incomplete PROJ pipelines

shawndove committed:
Merge branch 'main' into outsource_uncertainty
